### PR TITLE
Fix: improper return handling of rz_hex_str2bin()

### LINKED ISSUE #2014: Trailing nibble causes any disasm to fail
*If you would like to report a bug, please fill the template below*

### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| Architecture/bits of the file (mandatory)            | x86_64
| `rizin -v`  | rizin 0.1.0-git 25359 @ linux-x86-64 git.0.1.0-git commit: 8d81d392c76aea394019fd2993f5d32f3c1f94a2 build: 2021-01-03__19:10:40

### Expected behavior
`rz-asm -a x86 -b64 -d 0x55f`
```
push rbp
invalid
```

### Actual behavior
`rz-asm -a x86 -b64 -d 0x55f`
```
invalid
```
### Note
This is only the shortest failing demo. The same bug applies to any valid assembly byte sequence that has an additional trailing nibble.

### PR DESCRIPTION
**Pull request by PeiweiHu**
Fix https://github.com/rizinorg/rizin/issues/2014

**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

**Comment by PeiweiHu** [comment]
BTW, some other invoking parts also class the negative return value of rz_hex_str2bin() to error. Maybe they need to fix too. IMO, adding 0 to the tail doesn't mean absolute error. Maybe the default behavior sometimes.

**Review by ret2libc — CHANGES_REQUESTED** [review]
I'm not sure this is the proper fix for the problem because there are many other use cases of `rz_hex_str2bin` all around the codebase.

I think we should probably fix a bit the function itself so that it correctly accepts nibbles and it returns the right value for that, while returning a negative(error) value for really invalid strings.

**Comment by ret2libc** [comment]
If you see at `test_hex.c`, I would expect that the new `rz_hex_str2bin` function would pass all these tests:
```C
	mu_assert_eq(rz_hex_str2bin("41424344", buf), 4, "4 bytes are written");
	mu_assert_memeq(buf, (ut8 *)"ABCD", 4, "ABCD has been written");
	mu_assert_eq(rz_hex_str2bin("0x41424344", buf), 4, "4 bytes are written");
	mu_assert_memeq(buf, (ut8 *)"ABCD", 4, "ABCD has been written");
	mu_assert_eq(rz_hex_str2bin("616263646566", buf), 6, "6 bytes are written");
	mu_assert_memeq(buf, (ut8 *)"abcdef", 6, "abcdef has been written");
	mu_assert_eq(rz_hex_str2bin("61626364656", buf), 6, "last byte with a single nibble should be correctly returned");
	mu_assert_memeq(buf, (ut8 *)"abcde`", 6, "last nibble is interprted as 0x60");
	mu_assert_true(rz_hex_str2bin("hello", buf) < 0, "an error should be returned because invalid string");
	mu_assert_true(rz_hex_str2bin("hello123456", buf) < 0, "an error should be returned because invalid string");
	mu_assert_true(rz_hex_str2bin("123456hello", buf) < 0, "an error should be returned because invalid string");
```
Does it make sense?

**Comment by PeiweiHu** [comment]
> If you see at `test_hex.c`, I would expect that the new `rz_hex_str2bin` function would pass all these tests:
> 
> ```c
> 	mu_assert_eq(rz_hex_str2bin("41424344", buf), 4, "4 bytes are written");
> 	mu_assert_memeq(buf, (ut8 *)"ABCD", 4, "ABCD has been written");
> 	mu_assert_eq(rz_hex_str2bin("0x41424344", buf), 4, "4 bytes are written");
> 	mu_assert_memeq(buf, (ut8 *)"ABCD", 4, "ABCD has been written");
> 	mu_assert_eq(rz_hex_str2bin("616263646566", buf), 6, "6 bytes are written");
> 	mu_assert_memeq(buf, (ut8 *)"abcdef", 6, "abcdef has been written");
> 	mu_assert_eq(rz_hex_str2bin("61626364656", buf), 6, "last byte with a single nibble should be correctly returned");
> 	mu_assert_memeq(buf, (ut8 *)"abcde`", 6, "last nibble is interprted as 0x60");
> 	mu_assert_true(rz_hex_str2bin("hello", buf) < 0, "an error should be returned because invalid string");
> 	mu_assert_true(rz_hex_str2bin("hello123456", buf) < 0, "an error should be returned because invalid string");
> 	mu_assert_true(rz_hex_str2bin("123456hello", buf) < 0, "an error should be returned because invalid string");
> ```
> 
> Does it make sense?

Yep, I will try to refine rz_hex_str2bin.

**Comment by XVilka** [comment]
Two failed tests:
```
[XX] db/tools/rz_ax rz-ax -K
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc '!printf 1 | rz-ax -K | grep -c 0x00000001' =
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-1
+0


[XX] db/cmd/cmd_search /x wrong entries
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc '/x 653b2530000000724b53b834e7150883ec188' bins/elf/analysis/go_stripped
-- stdout
--- expected
+++ actual
@@ -1,2 +1,2 @@
-0x0805ae90 hit0_0 653b2530000000724b53b834e7150883ec
-0x0805b030 hit0_1 653b2530000000724b53b834e7150883ec
+0x0805ae90 hit0_0 653b2530000000724b53b834e7150883ec188b
+0x0805b030 hit0_1 653b2530000000724b53b834e7150883ec188b

-- stderr
Searching 19 bytes in [0x8235230-0x8235274]
```

**Comment by PeiweiHu** [comment]
> 

sorry for the bother. where can I get bins/elf/analysis/go_stripped?

**Comment by PeiweiHu** [comment]
> Two failed tests:
> 
> ```
> [XX] db/tools/rz_ax rz-ax -K
> RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc '!printf 1 | rz-ax -K | grep -c 0x00000001' =
> -- stdout
> --- expected
> +++ actual
> @@ -1,1 +1,1 @@
> -1
> +0
> 
> 
> [XX] db/cmd/cmd_search /x wrong entries
> RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc '/x 653b2530000000724b53b834e7150883ec188' bins/elf/analysis/go_stripped
> -- stdout
> --- expected
> +++ actual
> @@ -1,2 +1,2 @@
> -0x0805ae90 hit0_0 653b2530000000724b53b834e7150883ec
> -0x0805b030 hit0_1 653b2530000000724b53b834e7150883ec
> +0x0805ae90 hit0_0 653b2530000000724b53b834e7150883ec188b
> +0x0805b030 hit0_1 653b2530000000724b53b834e7150883ec188b
> 
> -- stderr
> Searching 19 bytes in [0x8235230-0x8235274]
> ```

sorry for the bother. where can I get bins/elf/analysis/go_stripped?

**Comment by XVilka** [comment]
@PeiweiHu see https://github.com/rizinorg/rizin-testbins

**Comment by XVilka** [comment]
I wonder if it makes sense to just substitute calls of these two functions with one new function, or just make it new `rz_hex_str2bin()`.

**Comment by PeiweiHu** [comment]
> 

The suggested new behavior of rz_hex_str2bin() (see changes in **test_hex.c** )  will add 0 to tail when there is a nibble. It returns a positive length just like no nibble exists. However, this would not always be the proper behavior. That's why some tests failed before. I think we need a function to identify whether the nibble exists. 

To fix  https://github.com/rizinorg/rizin/issues/2014, the conservative way is don't change rz_hex_str2bin() since many other invoking places depend on its original behavior. If rz_hex_str2bin() is changed, many adaptive changes are needed too to keep the original behavior of the software.

File: librz/util/hex.c [hunk 1]
```diff
@@ -462,12 +462,61 @@ RZ_API int rz_hex_str2bin(const char *in, ut8 *out) {
...
 	return nibbles / 2;
 }
 
+/**
+ * \brief Check whether the input string \p in has nibble
+ * 
+ * \param in Input string in hexadecimal form. An optional "0x" prefix may be present.
+ * \return 1 if the input string has nibble, 0 otherwise. -1 if not a valid string. 
+ */
+RZ_API int rz_hex_str_has_nibble(const char *in) {
```
Comment (ret2libc) [inline librz/util/hex.c]: just return `bool` and true/false.

File: librz/util/hex.c [hunk 2]
```diff
@@ -462,12 +462,61 @@ RZ_API int rz_hex_str2bin(const char *in, ut8 *out) {
...
+		}
+		/* ignore character */
+		in++;
+	}
+
+	if (nibbles % 2) {
+		return 1;
+	}
+
+	return 0;
```
Comment (ret2libc) [inline librz/util/hex.c]:
```suggestion
	return nibbles % 2;
```

**Comment by ret2libc** [comment]
```
[XX] db/cmd/cmd_search /x wrong entries
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc '/x 653b2530000000724b53b834e7150883ec188' bins/elf/analysis/go_stripped
-- stdout
--- expected
+++ actual
@@ -1,2 +1,2 @@
-0x0805ae90 hit0_0 653b2530000000724b53b834e7150883ec
-0x0805b030 hit0_1 653b2530000000724b53b834e7150883ec
+0x0805ae90 hit0_0 653b2530000000724b53b834e7150883ec188b
+0x0805b030 hit0_1 653b2530000000724b53b834e7150883ec188b
```

I feel like the new behaviour is actually correct. You search for `653b2530000000724b53b834e7150883ec188` and you get the strings  `653b2530000000724b53b834e7150883ec188b`. The old behaviour was missing the last few bits of the searched string.

```
[XX] db/tools/rz_ax rz-ax -K
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc '!printf 1 | rz-ax -K | grep -c 0x00000001' =
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-1
+0
```

I feel here the problem is that `rz-ax` is using `rz_hex_str2bin` when instead it should be using `strtol` or `rz_num_math`, because the number 1 is considered as a number and not as an hex string.

**Comment by PeiweiHu** [comment]
To keep the original behavior after rz_hex_str2bin() is changed, I looked through the places invoking it and modified them (that's why so many files are changed). 

I wonder if you would like to keep the changes. Or just change code as less as possible only if CI test can be satisfied, so I will roll back my commit to - **refine rz_hex_str2bin(), add rz_hex_str_has_nibble(), change two files to satisfy failed test** (or 1 file if new search behavior is reasonable).

**Comment by ret2libc** [comment]
I'm not sure we need that `nibble` thing, but we should rather make sure that all pieces that call hex2bin are actually dealing with an hexstring and not with the hex representation of a number (there is a slight difference IMHO).

**Pull request closed without merging**